### PR TITLE
[ROOT7] Use std::string instead of std::string_view

### PR DESCRIPTION
**Pull request by amadio**
The interfaces of ROOT's `string_view` and `std::string_view` are
different (e.g., no `to_string()` member function in `std::string_view`).

Reference: http://en.cppreference.com/w/cpp/header/string_view

### CODE REVIEW

**Comment by amadio** [comment]
@phsft-bot build just on slc6/gcc62 with flags -Droot7=ON -Dcxx14=ON

**Comment by dpiparo** [comment]
Hi,

I wonder if it wouldn't be perhaps better to implement a to_string method to guarantee compatibility with the stl string_view. This will be also immediately useful for the internals of the TDF...

**Comment by amadio** [comment]
Hi Danilo, the problem is that the standard `string_view` doesn't have the `.to_string()` method, so we cannot fix it by adding it in ROOT. We probably have to move to using the free function `std::to_string()` specialized for `string_view` in that case.

**Comment by bluehood** [comment]
Hi,
our implementation is also missing an `operator<<` overload, see ROOT-8896.

P.S.
I think the standard way to convert a `std::string_view` to a `std::string` is just to construct the latter from the former.

**Comment by amadio** [comment]
My reasoning for removing usage of `std::string_view` is that using `const std::string&` avoids copying memory and only requires C++11, while `std::string_view` requires C++17 or a backport. Moreover, along with the extra requirements of `std::string_view`, when it is passed by value it uses a pointer + length, which means possibly twice the size of just a reference. If we are not using the editing potential of `std::string_view`, or even worse, are just using it to actually make a copy of the data inside, then I see no real advantage in using it relative to just using a reference to `std::string` in the interface that basically does the same job better without adding the extra requirements of a higher C++ standard and/or a backport.

**Comment by dpiparo** [comment]
I see two advantages of string_view:
1. Forward compatibility
2. TStrings, const char*, std::string, std::string_view can be passed

**Comment by amadio** [comment]
Ok, as long as we can compile with GCC7 in C++17 mode, as the experiments are requesting, I don't have a strong opinion on this and we can keep using `std::string_view`. That said, should I try to make point 2 above work with `const std::string&` or do we just close this PR?

File: io/io/v7/inc/ROOT/TFile.hxx [hunk 1]
```diff
@@ -49,8 +49,7 @@ private:
 
   /// Serialize the object at address, using the object's TClass.
   //FIXME: what about `cl` "pointing" to a base class?
-  void WriteMemoryWithType(std::string_view name, const void *address,
-                           TClass *cl);
+  void WriteMemoryWithType(const std::string &name, const void *address, TClass *cl);
```
Comment (Axel-Naumann) [inline io/io/v7/inc/ROOT/TFile.hxx]: I'm okay with this one.

File: io/io/v7/inc/ROOT/TFile.hxx [hunk 2]
```diff
@@ -82,36 +81,32 @@ public:
...
-                           const Options_t &opts = Options_t());
+  static TFilePtr Recreate(const std::string &name, const Options_t &opts = Options_t());
 
   ///\}
 
   /// Set the new directory used for cached reads, returns the old directory.
   ///
   /// \note: Synchronizes multi-threaded accesses through locks.
-  static std::string SetCacheDir(std::string_view path);
+  static std::string SetCacheDir(const std::string &path);
```
Comment (Axel-Naumann) [inline io/io/v7/inc/ROOT/TFile.hxx]: Okay (as are the remaining changes below).

File: io/io/v7/src/TFile.cxx [hunk 3]
```diff
@@ -124,75 +122,78 @@ static std::mutex& GetCacheDirMutex() {
...
-    std::lock_guard<std::mutex> fLock;
-
-    SetCacheDirRAII_t(bool need): fLock(GetCacheDirMutex()) {
-      if (need)
-        fOldCacheDir = TFile::GetCacheFileDir();
-    }
+static std::unique_ptr<ROOT::Experimental::Internal::TFileStorageInterface> OpenV6TFile(
+   const std::string &name, const char *mode, const ROOT::Experimental::TFile::Options_t &opts)
+{
+   // Set and re-set the cache dir.
```
Comment (Axel-Naumann) [inline io/io/v7/src/TFile.cxx]: The formatting changes would ideally be in a separate commit.

File: io/io/v7/inc/ROOT/TFile.hxx (hunk 1 above)
Comment (bluehood) [inline io/io/v7/inc/ROOT/TFile.hxx]: Hi Axel,
out of curiosity -- and to educate myself -- why is a `const std::string&` fine here and not above? You foresee a different usage? I.e. never `WriteMemoryWithType("pooh")`?

File: io/io/v7/src/TFile.cxx (hunk 3 above)
Comment (amadio) [inline io/io/v7/src/TFile.cxx]: I applied clang-format to each commit. If you insist I can separate the clang-format changes, but why not commit in the right format directly?

As for the other comments, I will address them and push a new version soon.

File: io/io/v7/inc/ROOT/TFile.hxx (hunk 1 above)
Comment (Axel-Naumann) [inline io/io/v7/inc/ROOT/TFile.hxx]: If the interface uses the set of characters as null terminated string (by definition, not just because the current implementation happens to do that) then a `std::string` is fine. If we do string manipulation anyway (extract a path name, remove `"?CACHE=3M&CERT=..."` from a URL) then there's no point in forcing the parameter to be copied into a null-terminated `std::string`.

File: io/io/v7/src/TFile.cxx (hunk 3 above)
Comment (Axel-Naumann) [inline io/io/v7/src/TFile.cxx]: > why not commit in the right format directly?
Indeed, I will merge that branch of mine once I get to that.

**Comment by amadio** [comment]
Closing this, since @Axel-Naumann's changes to ROOT7's `TFile` make this PR obsolete.
